tensorflow/python/ops/special_math_ops.py

```diff
@@ -324,13 +324,43 @@ def sort_key(input_index, a):
     num_summed_elements = _total_size(t0_shape[-len(axes_to_sum):])
     new_shape = t0_shape[:len(preserved_axes)] + (num_broadcast_elements_t0,
                                                   num_summed_elements)
+    # Make new_shape valid, if invalid to be used in reshape
+    none_dims = [i for i, d in enumerate(new_shape)
```

Comment:
Good point.  I suspect the overhead is negligible, but I haven't verified this.

I would be fine with always using array_ops.shape to keep the code simple.  Alternatively, to keep the number of ops small, you could create a _get_shape helper that either returns array_ops.shape or returns a tuple of integers when possible.  You could also modify _total_size so that it returns an integer if given a tuple of integers, and uses math_ops.reduce_prod otherwise.

It seems like this would get the best of both worlds, but let me know if I've missed something.